src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2585 @@
+    __ cmpb(rax, Address(rcx, 0));
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 3:
+    small_hs_jmp_table[3] = __ pc();
+    __ movzwl(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpw(Address(rcx, 0), rax);
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 4: CASE 5:
+    small_hs_jmp_table[4] = __ pc();
+    small_hs_jmp_table[5] = __ pc();
+    __ movl(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpl(rax, Address(rcx, 0));
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 6:
+    small_hs_jmp_table[6] = __ pc();
+    __ movl(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpl(rax, Address(rcx, 0));
+    __ jne(L_top_loop_1);
+    __ movzbl(rax, Address(rbx, r15, Address::times_1, -0x18));
+    __ cmpb(rax, Address(r10, 0x5));
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 7:
+    small_hs_jmp_table[7] = __ pc();
+    __ movl(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpl(rax, Address(rcx, 0));
+    __ jne(L_top_loop_1);
+    __ movzwl(rax, Address(rbx, r15, Address::times_1, -0x18));
+    __ cmpw(Address(r10, 0x5), rax);
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 8 CASE 9:
+    small_hs_jmp_table[8] = __ pc();
+    small_hs_jmp_table[9] = __ pc();
+    __ movq(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpq(rax, Address(rcx, 0));
+    __ je(L_0x406019);
+
+    //  CASE 10:
+    small_hs_jmp_table[10] = __ pc();
+    __ movq(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpq(rax, Address(r10, 0x1));
+    __ jne(L_top_loop_1);
+    __ movzbl(rax, Address(r10, 0x9));
+    __ cmpb(Address(rbx, r15, Address::times_1, -0x14), rax);
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 11:
+    small_hs_jmp_table[11] = __ pc();
+    __ movq(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpq(rax, Address(r10, 0x1));
+    __ jne(L_top_loop_1);
+    __ movzwl(rax, Address(r10, 0x9));
+    __ cmpw(Address(rbx, r15, Address::times_1, -0x14), rax);
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 12:
+    small_hs_jmp_table[12] = __ pc();
+    __ movq(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpq(rax, Address(rcx, 0));
+    __ jne(L_top_loop_1);
+    __ movzwl(rax, Address(rbx, r15, Address::times_1, -0x14));
+    __ cmpw(Address(r10, 0x9), rax);
+    __ jne(L_top_loop_1);
+    __ movzbl(rax, Address(rbx, r15, Address::times_1, -0x12));
+    __ cmpb(rax, Address(r10, 0xb));
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
+
+    //  CASE 13:
+    small_hs_jmp_table[13] = __ pc();
+    __ movq(rax, Address(rbx, r15, Address::times_1, -0x1c));
+    __ cmpq(rax, Address(r10, 0x1));
+    __ jne(L_top_loop_1);
+    __ movl(rax, Address(r10, 0x9));
+    __ cmpl(Address(rbx, r15, Address::times_1, -0x14), rax);
+    __ jne(L_top_loop_1);
+    __ jmp(L_0x406019);
```

Comment:
For cases which are multiple of 4 bytes we can use VMASKMOVPS (conditional moves) and VPTEST.

Reply:
Not sure what you mean here.  Could you elaborate (although it may be moot after all the changes)?